### PR TITLE
Pr/conv transposed adam activations

### PR DESCRIPTION
**Pull request by j0nny16**
Hi,

I used PositNN for my bachelor's thesis, comparing posit and float training on MNIST and FashionMNIST with denoising autoencoders. In order to test my specific networks, I had to expand this project a little bit. I added a few things, most notably a transposed convolution layer and the Adam optimizer. I have also found and fixed a bug in the convolution layer for strides > 1.

I have split the work into three pull requests:
- this PR — bug fixes plus transposed convolution and Adam
- #B — optional loss scaling and a fix for the reported loss value
- #C — the layers I needed for the GAN (LeakyReLU, BatchNorm2d, BCELoss). Please note that while all new layers from this PR have been tested against pytorch (as have the other new layers) and found to work, all code was written by Claude Opus 5 and only superficially checked by me. This was mostly because I have a lot to do right now, but was nevertheless very curious on how posits would handle GAN networks.

This PR is the base, B and C both build on this one.

# The main fix: Conv2d backward at stride > 1
Conv2d::backward and Conv2d::gradient are wrong whenever the forward floor division is not exact, i.e. whenever (H + 2*padding - kernel) % stride != 0. That includes the common 28 -> 14 and 14 -> 7 stages built from kernel 3, stride 2, padding 1.

## Two separate problems:

Input gradient. The old path rotated the weight and ran a regular convolution2d over delta. For a 14x14 input it returns a 13x13 gradient, so the preceding layer receives a tensor of the wrong shape, and the border rows that the floor division drops are treated as contributing nothing even though those input positions do feed an output.

Weight gradient. convolution2d_gradient returns a kernel gradient up to stride-1 too large in each spatial dimension — 4x4 for a 3x3 kernel at stride 2. weight_gradient += temp then accumulates by flat index, silently mixing unrelated kernel positions.

I checked this against a direct reference implementation of the convolution gradients in double. On master, 6 of 12 checks fail and the weight gradient reaches a relative error of about 1.0, i.e. essentially unrelated to the correct value. After the fix all 12 agree to ~1e-8 in posit<32,2>. Geometries where the division is exact, and stride 1, were already correct and are unchanged.

The fix computes the input gradient as a transposed convolution of delta with the same weight, which is what it mathematically is, with output_padding chosen to reproduce the original input size. The weight gradient is cropped to the top-left kernel_h x kernel_w block.

## Two behaviour changes you should know about
These are deliberate, but they are not pure bug fixes, so please push back if you would rather have them separately or not at all:

Conv2d::gradient no longer divides by the batch size. No other layer does this, so the mean reduction was applied inconsistently depending on which layers a model contained. It now happens once, in MSELoss::derivative, where the requested Reduction is actually known. This is the same place PyTorch applies it.
Conv2d::reset_parameters now initializes the weights. It previously printed "Conv2d layer is not initialized" and left them at zero, so the layer could not be trained without initializing it by hand. It now uses the same kaiming_uniform as torch.nn.Conv2d, matching the new TransposedConv2d.

# New Components

## Transposed Conv
Uses basically the same loop as the normal convolution. The methods to build the windows are built directly into transposed_convolution.hpp and not into window itself.

## Adam
The moment updates go through fused(), following the SGD implementation.

# The rest
Sigmoid and Tanh no longer produce NaR for large inputs (> 709). Both evaluated exp() at a large positive argument for inputs far from zero; the double fallback returns inf and inf -> posit is NaR, which then spreads through the network. With posit<16,1> this reproducibly turned into NaR weights during training. Both are rewritten as a case split so the argument of exp() is always <= 0. This works around the problem rather than solving it. I think the real fix belongs in universal's exp(), which should saturate at maxpos instead of returning NaR when its double fallback overflows. I left a comment saying so at both call sites, since I did not want to patch a submodule from here.

# Scope
Only include/positnn/ is touched. examples/ and the universal submodule are untouched. The branch is split into five commits, each of which compiles on its own.

I am happy to split this differently, drop the behaviour changes, or adjust anything to fit how you would rather have it. If you want a more thorough explanation on anything, I am happy to provide a (translated) snippet from my thesis. Thanks for publishing the library -  it made my thesis possible :)

Also let me know, if you want the above mentioned PRs for loss scaling and/or the LLM-built layers for a GAN.

### CODE REVIEW

**Review by gonced8 — CHANGES_REQUESTED** [review]
I found five correctness issues affecting mixed-precision initialization, convolution initialization, MSE gradients, low-precision Adam, and rectangular Conv2d backward shapes. All five were reproduced with GCC 16.1.1 against head `098104b`.

Written by Codex in behalf of Gonçalo Raposo.

**Comment by j0nny16** [comment]
Hey, thanks for your quick response. I'll send over my thesis as soon as I have a satisfactory english translation going, which should not take too long given the help of LLMs.

As for the problems you outlined, with help of Claude I could reproduce all of them and submit fixes. Attached is a markdown file with detailed reponses to each of the issues: [review-response.md](https://github.com/user-attachments/files/31102014/review-response.md)

To confirm, here are the tests I ran: [review_findings_reproducer.cpp](https://github.com/user-attachments/files/31100917/review_findings_reproducer.cpp)

```bash
g++ -std=c++17 -O1 -w -DHL_THREADS=4 -DLL_THREADS=4 -I posit-neuralnet/include/universal/include -I posit-neuralnet/include review_findings_reproducer.cpp -o review && ./review
```

The first 3 findings were completely new to me, however the last two were known by me already. Regarding 4, I was basically just going with it, because I did not use as low as 8 bit posits as optimizer precision with these hyperparameters because of that. IIRC changing up the hyperparameters to fit with 8 bit posits did not yield sensible results and so I just switched to 16 bit optimizers. For number 5, I did just use square inputs (MNIST, FashionMNIST) so that problem never became apparent.

Do let me know what you think, and also if you are generally interested in the other 2 two PRs based off of this one.

File: include/positnn/optimizer/Adam.hpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        _beta1_t *= _beta_1;
+        _beta2_t *= _beta_2;
+        T const one(1);
+        _bias_corr1 = one / (one - _beta1_t);
+        _bias_corr2 = one / (one - _beta2_t);
+
+        Optimizer<T>::step();
+    }
+
+    AdamOptions<T>& options() { return _options; }
```
Comment (gonced8) [inline include/positnn/optimizer/Adam.hpp]: **High:** `options()` still returns a mutable reference, but the converted learning rate is now cached only in the constructor. Changing `optimizer.options().learning_rate` later therefore has no effect, unlike the original implementation and the existing SGD API. In a reproducer, setting the learning rate to zero before the first step still changed the weight from `1` to `0.98999`. Please synchronize and validate mutable options before each step, or replace the mutable options API with explicit setters that update the cached values.

Written by Codex in behalf of Gonçalo Raposo.

**Review by gonced8 — CHANGES_REQUESTED** [review]
The five original findings are fixed, but the Adam option-storage change introduces one follow-up regression in the mutable options API.

Written by Codex in behalf of Gonçalo Raposo.

**Comment by gonced8** [comment]
@j0nny16 Thank you for resolving those 5 things! The fixes look good!
I ran Codex one more time and it found another thing in the Adam optimizer. It should be fairly simple to fix. Do you think you could add it to this PR too please?

**Comment by j0nny16** [comment]
@gonced8 not a problem, I have pushed a fix.

For this PR I decided for the syncing option. This adds a bit of performance overhead to the optimizer, but given the emulated posits this can probably be neglected.
I think in a possible refactor unifying the different optimizers it would be advisable to change this to a setter/getter approach, but this would change the API and stop your examples from working. So syncing it is for now.

Claude also found two problems with SGD/SGDMixed, which I left untouched for now.

For details see [review-answer2.md](https://github.com/user-attachments/files/31116799/review-answer2.md). The mentioned verification script is [review_round2_mutable_options.cpp](https://github.com/user-attachments/files/31116752/review_round2_mutable_options.cpp). The followup verification script is [sgd_mutable_options_followup.cpp](https://github.com/user-attachments/files/31116757/sgd_mutable_options_followup.cpp).

**Comment by gonced8** [comment]
Thank you @j0nny16 ! I'll approve and merge it!

**Review by gonced8 — APPROVED** [review]
